Add initial scaffold for unit testing via doctest

Neverlord opened the pull request:
Now that we switched to object libraries in the repository via #674, I could revisit C++ unit testing via `doctest` (see #643).

What makes `doctest` different from other unit tests is that it merges test code and implementation. There is no separate unit test binary and `doctest` has the concept of running tests before entering the regular `main` code. However, I'm unsure whether this is a good default. I've opted to always return with `return context.run()` from `main` after running any test code, because I think a user either wants to run `zeek` or to run the unit tests.

Also, since this bakes the unit tests into the `zeek` binary, I think explicitly enabling unit testing as opt in makes more sense than an opt out build flag. Otherwise, users might end up accidentally using a `zeek` binary with builtin unit tests in production.

When building `zeek` with unit tests, passing `--testing` as first argument on the command line enters "doctest mode". By default, this runs all test cases. All arguments after `--testing` are passed to `doctest`, to `--testing -h` prints the `doctest`-generated help text. Users also can add regular Zeek arguments by adding `--` after all `doctest` options followed by the `zeek` options. Just in case some part of Zeek accesses `bro_argc`/`bro_argv` (or if we decide to not always stop after running unit tests).

I've added two simple unit tests to make sure everything's working as expected. The CMake scaffold also generates CTest targets, so users can use the familiar CMake work flows. Further, this makes it simple to integrate automatic testing into CI jobs.

jsiwek commented:
Thanks Dominik, generally sounds good to me with a few suggestions:

* I'd shorten `zeek --testing` to `zeek --test`
* CI is reporting some btests broken now, maybe due to new argument parsing, so I didn't review that part closely yet in case you had time to look into it (or just let me know if I can help out there)
* Maybe do argument parsing the same way regardless of configuration?  That way it can emit a nicer error like "Zeek has not been configured to include unit tests" if one tries `zeek --test` in a non-doctest build.
* Doesn't seem much downside to using  `-DDOCTEST_CONFIG_SUPER_FAST_ASSERTS` from the start to minimize compile times?  I'd make `./configure --enable-cpp-tests` do that by default and maybe make `./configure --enable-cpp-tests --disable-fast-cpp-tests` be the way to omit that option.

jsiwek commented:
Sorry just thought these, too:

* Add the `--test` option to the `usage()` info.
* Add doctest's license to the `COPYING.3rdparty` file

Neverlord commented:
> Doesn't seem much downside to using `-DDOCTEST_CONFIG_SUPER_FAST_ASSERTS` from the start to minimize compile times? I'd make `./configure --enable-cpp-tests` do that by default and maybe make `./configure --enable-cpp-tests --disable-fast-cpp-tests` be the way to omit that option.

Enabling `DOCTEST_CONFIG_SUPER_FAST_ASSERTS` turns off expression decomposition.

So instead of writing  `CHECK(a == b)`, we'll always have to use `CHECK_EQ(a, b)`. I personally prefer the latter style anyways and I've added the define to CMake as you've suggested. However, I don't see when you would want to disable this option. Either we pick the `CHECK_EQ`-style or we don't. 🙂

I've integrated all of your other feedback and added the doctest license to `COPYING.3rdparty`.

jsiwek commented:
> Enabling `DOCTEST_CONFIG_SUPER_FAST_ASSERTS` turns off expression decomposition.
> So instead of writing `CHECK(a == b)`, we'll always have to use `CHECK_EQ(a, b)`.

Got the impression from this documentation that you could still use that style of check with the super-fast option:

https://github.com/onqtam/doctest/blob/master/doc/markdown/benchmarks.md#cost-of-an-assertion-macro

> However, I don't see when you would want to disable this option. Either we pick the `CHECK_EQ`-style or we don't. 🙂

There was also this quote from the docs about the super-fast option: "the only downside of this is: if an assert fails and a debugger is attached - when it breaks it will be in an internal function - the user will have to go 1 level up in the callstack to see the actual assert."

But yeah, think it's ok to not complicate things with more `./configure` options unless someone comes along that wants to make use of those complications :)

Otherwise, all merged now, thanks.

Couple minor notes for later (just to put them out there):
* Exact place we run the tests might need to move, but can do that when need arises.  E.g. may need to initialize certain global state or parse scripts for some tests.
* I'll update the CI scripts to run the new unit tests

Neverlord commented:
> Got the impression from this documentation that you could still use that style of check with the super-fast option:

You're right, I've must've mixed something up.

Thanks for merging! Now I can go ahead and migrate the unit tests for `threading_val_to_data` that are still lying around on my disk. 👍
